Add option to run pillows with gevent

millerdev opened the pull request:
Adds a `--gevent-workers=N` option that spawns N gevent workers in non-migration processes (processes other than process number 0 when there is a dedicated migration process) on the `run_ptop` management command. The number of pillow processors can be calculated with the formula:
```py
gevent_workers * (num_processes - (1 if dedicated_migration_process else 0))
```

If there is no dedicated migration process, gevent worker number 0 in each process will attempt to run migrations. However, there is a shared (Redis) locking mechanism that allows multiple workers to coordinate migration activity, limiting the frequency of migrations and preventing duplicate work (see `TaskCoordinator` added in [an earlier PR](https://github.com/dimagi/commcare-hq/pull/36819)).

https://dimagi.atlassian.net/browse/SAAS-18388

:blowfish: Review by commit.

## Safety Assurance

### Safety story

Changes in this PR enable concurrent pillow processing within a single process. The code has been evaluated for concurrency safety, and at this time there are no known usages of shared resources, etc. that may cause concurrency issues. However, it is possible that something was overlooked, so QA testing will be carried out on staging to improve confidence in the solution.

### Automated test coverage

Yes.

### QA Plan

https://dimagi.atlassian.net/browse/QA-8068

### Rollback instructions

<!--
If this PR follows standards of revertability, check the box below.
Otherwise replace it with detailed instructions or reasons a rollback is impossible.
-->

- [x] This PR can be reverted after deploy with no further considerations

millerdev committed:
Add option to run pillows with gevent

millerdev committed:
Share adapter caches among gevent workers

millerdev committed:
Add gevent locking to TTLCache

millerdev commented:
Hold off on review until this is no longer a draft. I intend to force-push to fix tests.

kaapstorm reviewed (commented):
No concerns about the functionality. Just a request about the code.

gherceg reviewed (approved):
Agreed on changing the name of the locking method and the inner method to better distinguish, but also came around to the idea of not being able to call the "non-locking" method since it's name is effectively overwritten by the locking method.

millerdev committed:
Rename lock-decorated methods

millerdev committed:
Add process (aka worker) number to pillow logs

millerdev committed:
Add more information to change chunk logging

millerdev committed:
Add error handling for gevent workers

Pillow processes are automatically restarted by supervisor if they exit
because of an unexpected error. However, previously if an individual
gevent worker died, it would not be restarted immediately because the
process would continue to run while workers remained alive. Kafka
partitions assigned to the dead worker would stagnate, causing pillow lag.

millerdev committed:
Prevent database connection timeouts in pillows

Use request_finished signal to close connections because other things
such as custom SQLAlchemy connection management in
corehq.sql_db.connections also connect to and trigger cleanup
on the signal, but are not triggered by django.db.close_old_connections().

Do not close old connections when running tests because it disrupts
things like class cleanup.
